src/GraphControl/Logger/TraceLogger.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "TraceLogger.h"
+
+using namespace Concurrency;
+using namespace std;
+using namespace Platform;
+using namespace winrt;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Diagnostics;
+using namespace winrt::Windows::Globalization;
+
+namespace GraphControl
+{
+    static reader_writer_lock s_traceLoggerLock;
+
+    constexpr auto GraphingMode = L"Graphing";
+    constexpr auto SessionGuid = L"SessionGuid";
+    constexpr auto CalcMode = L"CalcMode";
+
+    // Diagnostics events. Uploaded to asimov.
+    constexpr auto EVENT_NAME_EQUATION_COUNT_CHANGED = L"EquationCountChanged";
+    constexpr auto EVENT_NAME_FUNCTION_ANALYSIS_PERFORMED = L"FunctionAnalysisPerformed";
+    constexpr auto EVENT_NAME_VARIABLES_ADDED = L"VariablesAdded";
+
+    constexpr auto PDT_PRIVACY_DATA_TAG = L"PartA_PrivTags";
+    constexpr auto PDT_PRODUCT_AND_SERVICE_USAGE = 0x0000'0000'0200'0000u;
+
+#ifdef SEND_DIAGNOSTICS
+    // c.f. WINEVENT_KEYWORD_RESERVED_63-56 0xFF00000000000000 // Bits 63-56 - channel keywords
+    // c.f. WINEVENT_KEYWORD_*              0x00FF000000000000 // Bits 55-48 - system-reserved keywords
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_1 = 0x0000800000000000; // Bit 47
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_2 = 0x0000400000000000; // Bit 46
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_3 = 0x0000200000000000; // Bit 45
+#else
+    // define all Keyword options as 0 when we do not want to upload app diagnostics
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_1 = 0;
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_2 = 0;
+    constexpr int64_t MICROSOFT_KEYWORD_LEVEL_3 = 0;
+#endif
+
+#pragma region TraceLogger setup and cleanup
+
+    TraceLogger::TraceLogger()
+        : g_calculatorProvider(
+            L"MicrosoftCalculator",
```

Comment:
Should we use a more explicit name to distinct these logs from https://github.com/microsoft/calculator/blob/577aafb3f4b405c5cf64444a3f9e8e147259bf8f/src/CalcViewModel/Common/TraceLogger.cpp#L61? "MicrosoftGraphingCalculator"?

Reply:
This name corresponds to the event name that is used in the diagnostics, I don't think we should change it because all of these events are from the calculator app.